### PR TITLE
Add an option to enable runtime compilation

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/dotnet/aspnetcore/issues/17272

## Description
In ASP.NET Core 3.0 we've moved the Runtime Compilation feature into a package and made the feature optional as part of the work to remove Roslyn dependencies from the API surface. After that we've hard too many requests from customers who were confused for not having the changes automatically visible on their projects.
To help with this, we are adding a new option for the Razor Project template which can enable runtime-compilation for the user, during project creation.

## Customer Impact
Developers no more can see their changes after refreshing the browser while debugging. They have to stop the debugger, rebuild and see the changes. We do have documentation in place about how to enable this feature if needed (by referencing the new package) but too many customers are confused.

## Risk
Low: this is a scoped addition and no existing functionality is being touched.

### CODE REVIEW

**Comment by pranavkm** [comment]
Feature in pictures

image

```
> dotnet new razor -o razor1 --help

  ...

  -rrc|--razor-runtime-compilation  Determines if the project is configured to use runtime compilation in Debug builds.
                                    bool - Optional
                                    Default: false / (*) true
```

image

image

**Comment by pranavkm** [comment]
FYI @DamianEdwards \ @rynowak \ @danroth27 

/cc @mkArtakMSFT this is patch consideration.

**Review by ryanbrandenburg — APPROVED** [review]

**Review by mkArtakMSFT — APPROVED** [review]

**Comment by DamianEdwards** [comment]
Are we at all concerned about the double adding in DEBUG builds? Are we 100% confident there aren't side effects of doing that? And even if there aren't, do we want that in our templates?

**Comment by nguerrera** [comment]
Conditional package references on configuration can be dicey since the assets file path is not usually configuration specific. Cc @nkolev92

**Comment by pranavkm** [comment]
Thanks @nguerrera. I hadn't thought of that. We were considering using PrivateAssets="All" instead + some runtime magic to light it up locally

File: src/SignalR/clients/ts/signalr-protocol-msgpack/package.json [hunk 1]
```diff
@@ -42,7 +42,7 @@
     "src/**/*"
   ],
   "dependencies": {
-    "@microsoft/signalr": "link:../signalr",
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr-protocol-msgpack/package.json]: ? no

**Comment by PillowMetal** [comment]
> Thanks @nguerrera. I hadn't thought of that. We were considering using PrivateAssets="All" instead + some runtime magic to light it up locally

@nguerrera I see that the final solution rolled out into the new template was to not use PrivateAssets="All" at all, and the runtime magic was to include a new variable in the launchsettings.json file.

However, I liked the former conditional debug approach better because then I didn't deploy the razor runtime compilation files on release builds to my test and production servers.

Now, the change in launchsettings.json forces the web.config to update as well, injects changes into the .deps.json file, which all gets deployed on publish regardless of build configuration,  and REQUIRES that the razor DLLs be present (which is probably why you couldn't use PrivateAssets="All")

Adding back the conditional inclusion of the package in the project file does not change this behavior, so I would like to go back to conditionally including the Add extension method in the startup file as previously recommended, and go back to not including these files in release builds.

Is the AddRazorRuntimeCompliation method going to remain and not get obsoleted or deprecated? I am not sure what people found so confusing about that approach in the first place TBH.

P.S. I am curious to know why runtime compilation worked and still works with no code or packages necessary at all, as long as you run from the IDE without debugging... now THAT's magic :P

**Comment by mkArtakMSFT** [comment]
@PillowMetal, you're commenting on a closed issue. Please file a new issue and share your feedback there, so that we can review it during triage.

**Pull request merged**